### ISSUE TITLE
Expanding the XHR interface to mirror expansions to the fetch interface?

### ISSUE THREAD

**Issue by davidvancleve** [whatwg/xhr#284]
If a new Web Platform feature is primarily an extension to Fetch (for instance, defining an extension to the API and some corresponding new request handling logic), does it seem in general like it might be reasonable to mirror the fetch API changes (e.g. expanding `RequestInit`) with corresponding XHR changes? The motivation would be that there are some people who, for whatever reason, can't use the `fetch` API but can use XHR; this would allow them to make use of the new Fetch-first Web Platform feature.

(The specific example I have in mind is the Trust Token API, which [expands](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/fetch/request_init.idl?q=request_init.idl&sq=package:chromium&g=0&l=1&dlp=chromium&dlf=src/third_party/blink/renderer/core/fetch/request_init.idl&dlc=491a1c39fb7d97c698af068a3ee5e7515fd84f3c&dlr=4&dlgp=third_party/blink/renderer/core/fetch/request_init.idl&dlgr=chromium/chromium/src&drp=chromium&drf=src/third_party/blink/renderer/core/fetch/request_init.idl&drc=fc9c78109a556c467c09e1fcec92fe156fd2a783&drr=5&drgp=third_party/blink/renderer/core/fetch/request_init.idl&drgr=chromium/chromium/src) `RequestInit` to take a new argument specifying a cryptographic operation to execute alongside the request; in its prototype implementation in Chromium, we [add](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/xmlhttprequest/xml_http_request.idl?q=xml_http_request.idl&sq=package:chromium&dr&dlp=chromium&dlf=src/third_party/blink/renderer/core/xmlhttprequest/xml_http_request.idl&dlc=c2e7d4f0b24814b0d1c51a964db34ec5b4930756&dlr=130&dlgp=third_party/blink/renderer/core/xmlhttprequest/xml_http_request.idl&dlgr=chromium/chromium/src&drp=chromium&drf=src/third_party/blink/renderer/core/xmlhttprequest/xml_http_request.idl&drc=c0228a98feda794aae87b52193cc7313bbf16a4a&drr=131&drgp=third_party/blink/renderer/core/xmlhttprequest/xml_http_request.idl&drgr=chromium/chromium/src) a corresponding new setter to `XMLHttpRequest` so clients can access the same functionality via XHR. @jeremyroman, who was reviewing my XHR change, suggested I drop by and ask whether there are any issues with this approach.)

**Comment by annevk** [comment]
No, we don't want to extend XMLHttpRequest with new features, except if it gets them through a shared primitive, but that would not be the case here. Also, could you please file an issue against whatwg/fetch with the proposed extension? (See https://annevankesteren.nl/2014/02/monkey-patch for context.)

**Comment by davidvancleve** [comment]
(cc @csharrison)

**Comment by davidvancleve** [comment]
At the risk of straying off the topic of _this_ issue, I don't understand what you'd like the whatwg/fetch issue to describe. Would it just be a high level "we have a prototype implementation including an extension to the API and an implicit extension to some of the spec's operations and seek to, at some point, codify this new functionality by extending the spec" kind of comment?

**Comment by annevk** [comment]
Yeah and maybe explicitly solicit feedback from the community on the approach taken.

**Comment by davidvancleve** [comment]
Sure; created https://github.com/whatwg/fetch/issues/1047.

**Comment by annevk** [comment]
Can this be closed or is this still being considered?

**Comment by davidvancleve** [comment]
Hi Anne, I'd like to ask that we keep this open for a couple weeks until
@dvorak42 uploads the draft Trust Tokens spec (which he's working on as a
local patch against Fetch), which might provide some motivation for
discussing this further (or might not, depending how it ends up).

BTW, one question: what's an example of a "shared primitive" in this
passage?

> we don't want to extend XMLHttpRequest with new features, except if it
gets them through a shared primitive

Thanks!

**Comment by annevk** [comment]
E.g., say that we started supporting fetching of `blah:` URLs. XMLHttpRequest would not go out of its way to exclude that. (On the other hand, we have excluded `ReadableStream` object support from XMLHttpRequest despite it initially supporting that: #277.)

**Comment by davidvancleve** [comment]
Got it! Thanks.
